File: 2nd year/Laboratories/MAP/7/src/main/java/utils/WeekUtils.java
package utils;

import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Calendar;
import java.util.Date;
import java.util.List;
import java.util.regex.Pattern;
import java.util.stream.Collectors;

public class WeekUtils {
    private List<Calendar> didacticWeeks = new ArrayList<>(28);

    public WeekUtils(String dateFormat, String dateSeparator, List<String> didacticPeriods) {
        SimpleDateFormat dateParser = new SimpleDateFormat(dateFormat);
        for (String didacticPeriod : didacticPeriods) {
            String interval = didacticPeriod.replaceAll("\\s+", "");
            try {
                Date intervalBeg = dateParser.parse(interval.split(Pattern.quote(dateSeparator))[0]);
                Date intervalEnd = dateParser.parse(interval.split(Pattern.quote(dateSeparator))[1]);
                Calendar i = Calendar.getInstance(), l = Calendar.getInstance();
                for (i.setTime(intervalBeg), l.setTime(intervalEnd); !i.after(l); i.add(Calendar.DATE, 7)) {
                    didacticWeeks.add((Calendar) i.clone());
                }
            } catch (ParseException e) {
                throw new IllegalArgumentException("didacticPeriods contains invalid dates or dateFormat is malformed");
            }
        }
        if (didacticWeeks.size() != 28) {
            throw new IllegalArgumentException("didacticPeriods should span a 28 weeks period");
        }
    }

    public Integer getCurrentDidacticWeek() {
        Calendar calendar = Calendar.getInstance();
        calendar.setTime(new Date());
        return getDidacticWeekForCalendar(calendar);
    }

    public Calendar getCalendarForDidacticWeek(Integer didacticWeek) {
        return (Calendar) didacticWeeks.get(didacticWeek - 1).clone();
    }

    private Integer getDidacticWeekForCalendar(Calendar calendar) {
        Integer calendarWeek = calendar.get(Calendar.WEEK_OF_YEAR);
        List<Integer> didacticCalendarWeeks = didacticWeeks.stream()
                .mapToInt(c -> c.get(Calendar.WEEK_OF_YEAR))
                .boxed()
                .collect(Collectors.toList());
        if (didacticCalendarWeeks.indexOf(calendarWeek) != -1) {
            return didacticCalendarWeeks.indexOf(calendarWeek) + 1;
        }
        return 0;
    }
}

File: 2nd year/Laboratories/MAP/L9/CRUD/Utils/WeekUtils.cs
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Text.RegularExpressions;

namespace CRUD.Utils
{
    public class WeekUtils
    {
        public const int SemesterWeekSpan = 14;

        private readonly CultureInfo _cultureInfo;
        private readonly DateTime[] _didacticWeeks = new DateTime[SemesterWeekSpan];

        public WeekUtils(string locale, char dateSeparator, IEnumerable<string> didacticPeriods)
        {
            _cultureInfo = new CultureInfo(locale);

            var currentWeek = 0;
            foreach (var didacticPeriod in didacticPeriods)
            {
                var period = Regex.Replace(didacticPeriod, "\\s+", "");
                var intervalBeg = DateTime.Parse(period.Split(dateSeparator)[0], _cultureInfo);
                var intervalEnd = DateTime.Parse(period.Split(dateSeparator)[1], _cultureInfo);

                for (var i = intervalBeg; i.CompareTo(intervalEnd) <= 0; i = i.AddDays(7))
                {
                    if (currentWeek >= SemesterWeekSpan)
                    {
                        throw new ArgumentException(nameof(didacticPeriods) + $" must span {SemesterWeekSpan} weeks");
                    }

                    _didacticWeeks[currentWeek++] = i;
                }
            }

            if (currentWeek < SemesterWeekSpan)
            {
                throw new ArgumentException(nameof(didacticPeriods) + $" must span {SemesterWeekSpan} weeks");
            }
        }

        public int GetCurrentDidacticWeek()
        {
            return GetDidacticWeekForDateTime(DateTime.Today);
        }

//        private DateTime GetDateTimeForDidacticWeek(int didacticWeek)
//        {
//            return _didacticWeeks[didacticWeek - 1];
//        }

        private int GetDidacticWeekForDateTime(DateTime dateTime)
        {
            var weekRule = _cultureInfo.DateTimeFormat.CalendarWeekRule;
            var firstDayOfWeek = _cultureInfo.DateTimeFormat.FirstDayOfWeek;

            var calendarWeek = _cultureInfo.Calendar.GetWeekOfYear(dateTime, weekRule, firstDayOfWeek);
            var didacticCalendarWeeks = _didacticWeeks
                .Select(time => _cultureInfo.Calendar.GetWeekOfYear(time, weekRule, firstDayOfWeek))
                .ToArray();

            if (didacticCalendarWeeks.Contains(calendarWeek))
            {
                return Array.IndexOf(didacticCalendarWeeks, calendarWeek) + 1;
            }

            return 0;
        }
    }
}